File: src/gwt/src/org/rstudio/core/client/widget/ModalDialogTracker.java
/*
 * ModalDialogTracker.java
 *
 * Copyright (C) 2022 by Posit Software, PBC
 *
 * Unless you have received this program directly from Posit Software pursuant
 * to the terms of a commercial license agreement with Posit Software, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */
package org.rstudio.core.client.widget;

import java.util.ArrayList;
import java.util.List;

import org.rstudio.core.client.BrowseCap;
import org.rstudio.core.client.Debug;
import org.rstudio.core.client.ElementIds;
import org.rstudio.core.client.StringUtil;
import org.rstudio.core.client.a11y.A11y;
import org.rstudio.core.client.command.impl.DesktopMenuCallback;
import org.rstudio.core.client.dom.DomUtils;
import org.rstudio.studio.client.application.Desktop;
import org.rstudio.studio.client.application.events.AriaLiveStatusEvent.Severity;

import com.google.gwt.dom.client.Element;
import com.google.gwt.dom.client.Style;
import com.google.gwt.user.client.DOM;
import com.google.gwt.user.client.ui.Panel;
import com.google.gwt.user.client.ui.PopupPanel;

public class ModalDialogTracker
{
   public static void onShow(PopupPanel panel)
   {
      ensureVisible(panel);
      dialogStack_.add(panel);

      if (Desktop.hasDesktopFrame()) {
         if (BrowseCap.isElectron())
            Desktop.getFrame().setNumGwtModalsShowing(numModalsShowing());
         DesktopMenuCallback.setMainMenuEnabled(false);
      }

      updateInert(true);
   }

   public static boolean isTopMost(PopupPanel panel)
   {
      return !dialogStack_.isEmpty() &&
             dialogStack_.get(dialogStack_.size() - 1) == panel;
   }

   public static void onHide(PopupPanel panel)
   {
      dialogStack_.removeIf(panel::equals);
      
      if (Desktop.hasDesktopFrame()) {
         int numModals = numModalsShowing();
         if (BrowseCap.isElectron())
            Desktop.getFrame().setNumGwtModalsShowing(numModals);
         if (numModals == 0)
            DesktopMenuCallback.setMainMenuEnabled(true);
      }
      updateInert(false);
   }

   public static int numModalsShowing()
   {
      return dialogStack_.size();
   }

   /**
    * Attempt to dispatch an aria-live message to the topmost popup that supports it, if any.
    * 
    * @param message message to announce
    * @param delayMs delay before announcing
    * @param severity polite or assertive announcement
    * @return true if message was dispatched, false if no popups support status messages
    */
   public static boolean dispatchAriaLiveStatus(String message, int delayMs, Severity severity)
   {
      for (int i = dialogStack_.size() - 1; i >= 0; i--)
      {
         if (dialogStack_.get(i) instanceof AriaLiveStatusReporter)
         {
            ((AriaLiveStatusReporter) dialogStack_.get(i)).reportStatus(message, delayMs, severity);
            return true;
         }
      }
      return false;
   }
   
   public static List<PopupPanel> getModalDialogs()
   {
      return dialogStack_;
   }

   private static void updateInert(boolean addedPopup)
   {
      assert(numModalsShowing() >= 0);

      if (numModalsShowing() == 0)
         setInertMainWindow(false); // closed last popup, reenable main window
      else if (numModalsShowing() == 1 && addedPopup)
         setInertMainWindow(true); // added first popup, disable main window
      else if (addedPopup)
         A11y.setInert(getSecondTopMostElement(), true); // disable previous topmost
      else
         A11y.setInert(getTopMostElement(), false); // enable newly exposed topmost
   }

   private static Element getTopMostElement()
   {
      if (dialogStack_.isEmpty())
         return null;
      return dialogStack_.get(dialogStack_.size() - 1).getElement();
   }

   private static Element getSecondTopMostElement()
   {
      if (dialogStack_.size() < 2)
         return null;
      return dialogStack_.get(dialogStack_.size() - 2).getElement();
   }

   private static void setInertMainWindow(boolean inert)
   {
      Element ideWrapper = DOM.getElementById("rstudio_container");
      Element ideFrame = DOM.getElementById("rstudio");
      Element satWrapper = DOM.getElementById(ElementIds.getElementId(ElementIds.SATELLITE_PANEL));

      if (ideWrapper != null)
         A11y.setInert(ideWrapper, inert);
      if (ideFrame != null)
         A11y.setInert(ideFrame, inert);
      if (satWrapper != null)
         A11y.setInert(satWrapper, inert);
   }

   private static void ensureVisible(Panel panel)
   {
      try
      {
         ensureVisibleImpl(panel);
      }
      catch (Exception e)
      {
         Debug.logException(e);
      }
   }

   private static void ensureVisibleImpl(Panel panel)
   {
      Style styles = DomUtils.getComputedStyles(panel.getElement());
      Integer zIndex = StringUtil.parseInt(styles.getZIndex(), 1000);
      panel.getElement().getStyle().setZIndex(zIndex + dialogStack_.size());
   }

   private static final List<PopupPanel> dialogStack_ = new ArrayList<>();
}

File: src/node/desktop/src/main/modal-dialog-tracker.ts
/*
 * modal-dialog-tracker.ts
 *
 * Copyright (C) 2023 by Posit Software, PBC
 *
 * Unless you have received this program directly from Posit Software pursuant
 * to the terms of a commercial license agreement with Posit Software, then
 * this program is licensed to you under the terms of version 3 of the
 * GNU Affero General Public License. This program is distributed WITHOUT
 * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
 * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
 * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
 *
 */

import { ModalDialog } from '../ui/modal-dialog';
import { appState } from './app-state';

/**
 * Tracks modals that should enable/disable the main menu, including:
 * - ModalDialog (src/node/desktop/src/ui/modal-dialog.ts) - eg. Choose R
 * - GWT modals - eg. Global Options
 * - Electron.dialog (native modals) - eg. dialog.showMessageBox(...)
 *
 * Based on src/gwt/src/org/rstudio/core/client/widget/ModalDialogTracker.java
 */
export class ModalDialogTracker {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private modals: ModalDialog<any>[] = [];
  private electronModalsShowing = 0;
  private gwtModalsShowing = 0;

  /**
   * Returns the number of modals showing across all modal types
   * @returns Number of modals showing across all modal types
   */
  public numModalsShowing(): number {
    return this.modals.length + this.gwtModalsShowing + this.electronModalsShowing;
  }

  /**
   * Re-enables the main menu if there are no modals showing
   */
  public maybeReenableMainMenu() {
    if (this.numModalsShowing() === 0) {
      appState().gwtCallback?.mainWindow.menuCallback.setMainMenuEnabled(true);
    }
  }

  /**
   * Adds a ModalDialog to tracking and disables the main menu
   * @param modal ModalDialog to add to tracking
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public async addModalDialog(modal: ModalDialog<any>) {
    this.modals.push(modal);
    appState().gwtCallback?.mainWindow.menuCallback.setMainMenuEnabled(false);
  }

  /**
   * Removes a ModalDialog from tracking and re-enables the main menu if applicable
   * @param modal ModalDialog to remove from tracking
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  public async removeModalDialog(modal: ModalDialog<any>) {
    this.modals = this.modals.filter((m) => m !== modal);
    this.maybeReenableMainMenu();
  }

  /**
   * Sets the number of GWT modals showing
   * @param gwtModalsShowing Number of GWT modals showing
   */
  public setNumGwtModalsShowing(gwtModalsShowing: number) {
    this.gwtModalsShowing = gwtModalsShowing;
  }

  /**
   * Tracks a modal dialog that is shown by Electron.dialog (async) to ensure that the
   * main menu is disabled while the dialog is open and re-enabled when the
   * dialog is closed.
   * @param func async Electron.dialog function to call
   * @returns Promise that resolves to the return value of func
   */
  public async trackElectronModalAsync<T>(func: () => Promise<T>): Promise<T> {
    this.addElectronModal();
    return func().finally(() => this.removeElectronModal());
  }

  /**
   * Tracks a modal dialog that is shown by Electron.dialog (sync) to ensure that the
   * main menu is disabled while the dialog is open and re-enabled when the
   * dialog is closed.
   * @param func Electron.dialog function to call
   * @returns Return value of func
   */
  public trackElectronModalSync<T>(func: () => T): T {
    this.addElectronModal();
    const retVal = func();
    this.removeElectronModal();
    return retVal;
  }

  /**
   * Increments the number of Electron modals showing and disables the main menu
   */
  private addElectronModal() {
    this.electronModalsShowing++;
    appState().gwtCallback?.mainWindow.menuCallback.setMainMenuEnabled(false);
  }

  /**
   * Decrements the number of Electron modals showing and re-enables the main menu
   */
  private removeElectronModal() {
    if (this.electronModalsShowing > 0) this.electronModalsShowing--;
    this.maybeReenableMainMenu();
  }
}
